lib/puppet/type/exec.rb

```diff
@@ -581,7 +581,16 @@ def check_all_attributes(refreshing = false)
           val = @parameters[check].value
           val = [val] unless val.is_a? Array
           val.each do |value|
-            return false unless @parameters[check].check(value)
+            if !@parameters[check].check(value)
+              # Give a debug message so users can figure out what command would have been
+              # but don't print sensitive commands or parameters in the clear
+              cmdstring = @parameters[:command].sensitive ? "[command redacted]" : @parameters[:command].value
+              checkstring = @parameters[check].sensitive ? "[redacted]" : @parameters[check].value
+
+              debug(_("'%{cmd}' won't be executed because of failed check %{check}: '%{checkvalue}'") % { cmd: cmdstring, check: check, checkvalue: checkstring })
```

Comment:
The `debug` method is mixed into the `Puppet::Type::Exec` class, so the resource title is automatically prefixed to the mesasge, which makes the `'[command redacted]'` redundant:

```
Debug: /Stage[main]/Main/Exec[example]: '[command redacted]' won't be executed because of failed check onlyif: '/usr/bin/false'
```
 Maybe shorten so it's

```
Debug: /Stage[main]/Main/Exec[example]: Execution skipped because 'onlyif' check failed: '/usr/bin/false'
```

Reply:
Note that the title and the command itself may not be the same.  Should the title also be checked for sensitive?  Should I have a different message if the title and command are the same?  The ticket itself is about adding in the command if the command is not executed, because otherwise it is not printed elsewhere.

Reply:
Yeah `title` is just the label/display name for the resource. If you don't care about the command being sensitive, you can specify the command to run as the title. But it gets verbose as the title shows up in reports, etc. So people usually specify a title and put the actual command in the `command` parameter. And you have to do that if the command is sensitive, as titles cannot be.

For this ticket, I was confused earlier. I thought the goal was to emit the `onlyif`/`unless` command, but I see we already do that, and the goal is to output the `command` that would be executed if it wasn't a noop run:

```
$ bundle exec puppet apply manifest.pp --noop --debug
...
Debug: Exec[example](provider=posix): Executing check '[redacted]'
Debug: Executing: '[redacted]'
Debug: /Stage[main]/Main/Exec[example]: '[command redacted]' won't be executed because of failed check onlyif: '[redacted]'
```

So for this PR, I'm thinking we should drop the `: '%{checkvalue}'` part (so we don't repeat the check command that we already executed).

Also I'm seeing that if you don't specify wrap the value to `command`, `unless`, `onlyif` as sensitive, then puppet emits the plaintext value, which was initially surprising to me:

```puppet
exec { 'example':
  command => '/usr/bin/true',
  onlyif => '/usr/bin/false',
}
```

```
$ bx puppet apply manifest.pp --noop --debug
...
Debug: Exec[example](provider=posix): Executing check '/usr/bin/false'
Debug: Executing: '/usr/bin/false'
Debug: /Stage[main]/Main/Exec[example]: '/usr/bin/true' won't be executed because of failed check onlyif: '/usr/bin/false'
```

But I think it makes sense. Otherwise we'd always redact the `command`, and there'd be no point to this ticket.

Reply:
Yes, the goal is to show the command if it is not executed, otherwise it never shows up:

```
% bx puppet apply --debug -e "exec { 'example': command => '/usr/bin/uname', unless => '/usr/bin/true', }" | grep uname
Debug: /Stage[main]/Main/Exec[example]: '/usr/bin/uname' won't be executed because of failed check 'unless'
```
I'm not sure this change is directly impacted by --noop.